<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Demo: Transcriptional Patterns in CD4 T Cells &#8212; Hotspot 0.9.0 documentation</title>
    <link rel="stylesheet" href="_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/my-styles.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Function Reference" href="hotspot.html" />
    <link rel="prev" title="Demo: Lineage-recorded data from Chan et al, (Nature 2019)" href="Lineage_Tutorial.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<script type="text/javascript" src="_static/js/jquery-1.11.0.min.js "></script>
<script type="text/javascript" src="_static/js/jquery-fix.js "></script>
<script type="text/javascript" src="_static/bootstrap-3.3.7/js/bootstrap.min.js "></script>
<script type="text/javascript" src="_static/bootstrap-sphinx.js "></script>

  </head><body>

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="index.html">
          Hotspot</a>
        <span class="navbar-text navbar-version pull-left"><b>0.9.0</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="hotspot.html">API</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="index.html">Tutorials <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="Spatial_Tutorial.html">Demo: Spatial data from Slide-seq</a></li>
<li class="toctree-l1"><a class="reference internal" href="Lineage_Tutorial.html">Demo: Lineage-recorded data from Chan et al, (Nature 2019)</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Demo: Transcriptional Patterns in CD4 T Cells</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">This Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">Demo: Transcriptional Patterns in CD4 T Cells</a><ul>
<li><a class="reference internal" href="#Covered-Here">Covered Here</a></li>
<li><a class="reference internal" href="#Load-data-from-h5-file-and-extract-CD4-T-cells">Load data from h5 file and extract CD4 T cells</a><ul>
<li><a class="reference internal" href="#Extract-the-CD4-T-Cells-based-on-surface-protein-abundance">Extract the CD4 T Cells based on surface protein abundance</a></li>
<li><a class="reference internal" href="#Compute-a-Latent-Space-with-PCA">Compute a Latent Space with PCA</a></li>
</ul>
</li>
<li><a class="reference internal" href="#Creating-the-Hotspot-object">Creating the Hotspot object</a></li>
<li><a class="reference internal" href="#Determining-informative-genes">Determining informative genes</a></li>
<li><a class="reference internal" href="#Grouping-genes-into-lineage-based-modules">Grouping genes into lineage-based modules</a></li>
<li><a class="reference internal" href="#Plotting-module-correlations">Plotting module correlations</a></li>
<li><a class="reference internal" href="#Summary-Module-Scores">Summary Module Scores</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="body col-md-12 content" role="main">
      
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container,
div.nbinput.container div.prompt,
div.nbinput.container div.input_area,
div.nbinput.container div[class*=highlight],
div.nbinput.container div[class*=highlight] pre,
div.nboutput.container,
div.nboutput.container div.prompt,
div.nboutput.container div.output_area,
div.nboutput.container div[class*=highlight],
div.nboutput.container div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    min-width: 7ex;
    padding-top: 0.4em;
    padding-right: 0.4em;
    text-align: right;
    flex: 0;
}
@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    padding: 0.4em;
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    background: #f5f5f5;
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<div class="section" id="Demo:-Transcriptional-Patterns-in-CD4-T-Cells">
<h1>Demo: Transcriptional Patterns in CD4 T Cells<a class="headerlink" href="#Demo:-Transcriptional-Patterns-in-CD4-T-Cells" title="Permalink to this headline">¶</a></h1>
<p>This demo uses public data from 10x Genomics</p>
<p><em>5k Peripheral blood mononuclear cells (PBMCs) from a healthy donor with cell surface proteins (Next GEM)</em></p>
<ul class="simple">
<li>as processed by Cell Ranger 3.1.0</li>
<li><a class="reference external" href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.1.0/5k_pbmc_protein_v3_nextgem">Dataset Link</a></li>
</ul>
<p>For this demo, you need the filtered h5 file, “Feature / cell matrix HDF5 (filtered)”, with the filename:</p>
<ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">5k_pbmc_protein_v3_filtered_feature_bc_matrix.h5</span></code></li>
</ul>
<div class="section" id="Covered-Here">
<h2>Covered Here<a class="headerlink" href="#Covered-Here" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>Data preprocessing and filtering</li>
<li>Computing Autocorrelation in Hotspot to identify lineage-related genes</li>
<li>Computing local correlations between lineage genes to identify heritable modules</li>
<li>Plotting modules, correlations, and module scores</li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">warnings</span><span class="p">;</span> <span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>

<span class="kn">import</span> <span class="nn">hotspot</span>
<span class="kn">import</span> <span class="nn">h5py</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">scipy.sparse</span> <span class="k">as</span> <span class="nn">sparse</span>

<span class="kn">from</span> <span class="nn">sklearn.decomposition</span> <span class="k">import</span> <span class="n">PCA</span>
<span class="kn">from</span> <span class="nn">umap</span> <span class="k">import</span> <span class="n">UMAP</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/data/yosef2/users/david.detomaso/programs/anaconda3/lib/python3.6/site-packages/sklearn/externals/joblib/__init__.py:15: DeprecationWarning: sklearn.externals.joblib is deprecated in 0.21 and will be removed in 0.23. Please import this functionality directly from joblib, which can be installed with: pip install joblib. If this warning is raised when loading pickled models, you may need to re-serialize those models with scikit-learn 0.21+.
  warnings.warn(msg, category=DeprecationWarning)
</pre></div></div>
</div>
</div>
<div class="section" id="Load-data-from-h5-file-and-extract-CD4-T-cells">
<h2>Load data from h5 file and extract CD4 T cells<a class="headerlink" href="#Load-data-from-h5-file-and-extract-CD4-T-cells" title="Permalink to this headline">¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">def</span> <span class="nf">load_data</span><span class="p">():</span>
    <span class="n">f</span> <span class="o">=</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="s2">&quot;5k_pbmc_protein_v3_filtered_feature_bc_matrix.h5&quot;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>

    <span class="n">barcodes</span> <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="s1">&#39;matrix/barcodes&#39;</span><span class="p">][:]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;str&#39;</span><span class="p">)</span>

    <span class="n">names</span> <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="s1">&#39;matrix/features/name&#39;</span><span class="p">][:]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;str&#39;</span><span class="p">)</span>
    <span class="n">ens_ids</span> <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="s1">&#39;matrix/features/id&#39;</span><span class="p">][:]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;str&#39;</span><span class="p">)</span>
    <span class="n">f_type</span> <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="s1">&#39;matrix/features/feature_type&#39;</span><span class="p">][:]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;str&#39;</span><span class="p">)</span>
    <span class="n">genome</span> <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="s1">&#39;matrix/features/genome&#39;</span><span class="p">][:]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;str&#39;</span><span class="p">)</span>
    <span class="n">pattern</span> <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="s1">&#39;matrix/features/id&#39;</span><span class="p">][:]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;str&#39;</span><span class="p">)</span>
    <span class="n">read</span> <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="s1">&#39;matrix/features/read&#39;</span><span class="p">][:]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;str&#39;</span><span class="p">)</span>
    <span class="n">sequence</span> <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="s1">&#39;matrix/features/sequence&#39;</span><span class="p">][:]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;str&#39;</span><span class="p">)</span>

    <span class="n">features</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
        <span class="s1">&#39;Name&#39;</span><span class="p">:</span> <span class="n">names</span><span class="p">,</span>
        <span class="s1">&#39;Id&#39;</span><span class="p">:</span> <span class="n">ens_ids</span><span class="p">,</span>
        <span class="s1">&#39;Type&#39;</span><span class="p">:</span> <span class="n">f_type</span><span class="p">,</span>
        <span class="s1">&#39;Genome&#39;</span><span class="p">:</span> <span class="n">genome</span><span class="p">,</span>
        <span class="s1">&#39;Pattern&#39;</span><span class="p">:</span> <span class="n">pattern</span><span class="p">,</span>
        <span class="s1">&#39;Read&#39;</span><span class="p">:</span> <span class="n">read</span><span class="p">,</span>
        <span class="s1">&#39;Sequence&#39;</span><span class="p">:</span> <span class="n">sequence</span><span class="p">,</span>
    <span class="p">})</span>

    <span class="n">data</span> <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="s1">&#39;matrix/data&#39;</span><span class="p">][:]</span>
    <span class="n">indices</span> <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="s1">&#39;matrix/indices&#39;</span><span class="p">][:]</span>
    <span class="n">indptr</span> <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="s1">&#39;matrix/indptr&#39;</span><span class="p">][:]</span>
    <span class="n">shape</span> <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="s1">&#39;matrix/shape&#39;</span><span class="p">][:]</span>


    <span class="n">matrix</span> <span class="o">=</span> <span class="n">sparse</span><span class="o">.</span><span class="n">csc_matrix</span><span class="p">(</span>
        <span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">indices</span><span class="p">,</span> <span class="n">indptr</span><span class="p">),</span>
        <span class="n">shape</span><span class="o">=</span><span class="n">shape</span>
    <span class="p">)</span>

    <span class="c1"># Split the matrix into the mRNA and the antibody counts</span>
    <span class="n">mRNA_rows</span> <span class="o">=</span> <span class="n">features</span><span class="o">.</span><span class="n">Type</span> <span class="o">==</span> <span class="s2">&quot;Gene Expression&quot;</span>
    <span class="n">i_mRNA_rows</span> <span class="o">=</span> <span class="n">mRNA_rows</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">nonzero</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">protein_rows</span> <span class="o">=</span> <span class="n">features</span><span class="o">.</span><span class="n">Type</span> <span class="o">==</span> <span class="s2">&quot;Antibody Capture&quot;</span>
    <span class="n">i_protein_rows</span> <span class="o">=</span> <span class="n">protein_rows</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">nonzero</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>

    <span class="n">proteins</span> <span class="o">=</span> <span class="n">matrix</span><span class="p">[</span><span class="n">i_protein_rows</span><span class="p">,</span> <span class="p">:]</span>
    <span class="n">mRNA</span> <span class="o">=</span> <span class="n">matrix</span><span class="p">[</span><span class="n">i_mRNA_rows</span><span class="p">,</span> <span class="p">:]</span>

    <span class="k">del</span> <span class="n">matrix</span>

    <span class="n">features_genes</span> <span class="o">=</span> <span class="n">features</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">mRNA_rows</span><span class="p">]</span>
    <span class="n">features_proteins</span> <span class="o">=</span> <span class="n">features</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">protein_rows</span><span class="p">]</span>

    <span class="k">del</span> <span class="n">features</span>

    <span class="c1"># Gene filtering</span>
    <span class="n">num_umi</span> <span class="o">=</span> <span class="n">mRNA</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">num_umi</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">num_umi</span><span class="p">)</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>

    <span class="n">num_umi_mat</span> <span class="o">=</span> <span class="n">sparse</span><span class="o">.</span><span class="n">lil_matrix</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">num_umi</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">num_umi</span><span class="p">)))</span>
    <span class="n">num_umi_mat</span><span class="o">.</span><span class="n">setdiag</span><span class="p">(</span><span class="mi">1</span><span class="o">/</span><span class="n">num_umi</span><span class="p">)</span>

    <span class="n">mito_genes</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;mt-&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">features_genes</span><span class="o">.</span><span class="n">Name</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;str&#39;</span><span class="p">)]</span>
    <span class="n">mito_genes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mito_genes</span><span class="p">)</span>

    <span class="n">mito_count</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mRNA</span><span class="p">[</span><span class="n">mito_genes</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">))</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
    <span class="n">mito_percent</span> <span class="o">=</span> <span class="n">mito_count</span> <span class="o">/</span> <span class="n">num_umi</span> <span class="o">*</span> <span class="mi">100</span>

    <span class="n">proteins_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
        <span class="n">proteins</span><span class="o">.</span><span class="n">toarray</span><span class="p">(),</span> <span class="n">index</span><span class="o">=</span><span class="n">features_proteins</span><span class="o">.</span><span class="n">Id</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;str&#39;</span><span class="p">),</span>
        <span class="n">columns</span><span class="o">=</span><span class="n">barcodes</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;str&#39;</span><span class="p">)</span>
    <span class="p">)</span><span class="o">.</span><span class="n">T</span>

    <span class="n">features_genes</span> <span class="o">=</span> <span class="n">features_genes</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;Id&quot;</span><span class="p">)[[</span><span class="s2">&quot;Name&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">rename</span><span class="p">({</span><span class="s2">&quot;Name&quot;</span><span class="p">:</span> <span class="s2">&quot;GeneSymbol&quot;</span><span class="p">},</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="n">num_umi</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">num_umi</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">barcodes</span><span class="p">)</span>
    <span class="n">mito_percent</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">mito_percent</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">barcodes</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">mRNA</span><span class="p">,</span> <span class="n">features_genes</span><span class="p">,</span> <span class="n">barcodes</span><span class="p">,</span> <span class="n">num_umi</span><span class="p">,</span> <span class="n">mito_percent</span><span class="p">,</span> <span class="n">proteins_df</span>

<span class="n">mRNA</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">barcodes</span><span class="p">,</span> <span class="n">umi_counts</span><span class="p">,</span> <span class="n">mito_percent</span><span class="p">,</span> <span class="n">proteins</span> <span class="o">=</span> <span class="n">load_data</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="section" id="Extract-the-CD4-T-Cells-based-on-surface-protein-abundance">
<h3>Extract the CD4 T Cells based on surface protein abundance<a class="headerlink" href="#Extract-the-CD4-T-Cells-based-on-surface-protein-abundance" title="Permalink to this headline">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">def</span> <span class="nf">clr</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>

    <span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">x</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">x</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">subtract</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">x</span>

<span class="n">proteins_clr</span> <span class="o">=</span> <span class="n">clr</span><span class="p">(</span><span class="n">proteins</span><span class="p">)</span>

<span class="n">is_cd4</span> <span class="o">=</span> <span class="p">(</span>
    <span class="p">(</span><span class="n">proteins_clr</span><span class="p">[</span><span class="s1">&#39;CD14&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">2</span><span class="p">)</span> <span class="o">&amp;</span>
    <span class="p">(</span><span class="n">proteins_clr</span><span class="p">[</span><span class="s1">&#39;CD4&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">3</span><span class="p">)</span> <span class="o">&amp;</span>
    <span class="p">(</span><span class="n">proteins_clr</span><span class="p">[</span><span class="s1">&#39;CD3&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mf">2.5</span><span class="p">)</span>
<span class="p">)</span><span class="o">.</span><span class="n">values</span>

<span class="c1"># Further filter by Mito-Percent and Num-Umi</span>
<span class="n">is_healthy</span> <span class="o">=</span> <span class="p">(</span>
    <span class="p">(</span><span class="n">umi_counts</span> <span class="o">&gt;</span> <span class="mi">3100</span><span class="p">)</span> <span class="o">&amp;</span>
    <span class="p">(</span><span class="n">mito_percent</span> <span class="o">&lt;</span> <span class="mi">16</span><span class="p">)</span>
<span class="p">)</span>

<span class="n">is_cd4</span> <span class="o">=</span> <span class="n">is_cd4</span> <span class="o">&amp;</span> <span class="n">is_healthy</span>
<span class="n">is_cd4_ii</span> <span class="o">=</span> <span class="n">is_cd4</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">nonzero</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>

<span class="n">mRNA_cd4</span> <span class="o">=</span> <span class="n">mRNA</span><span class="p">[:,</span> <span class="n">is_cd4_ii</span><span class="p">]</span>

<span class="c1"># Filter genes</span>
<span class="n">gene_counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">((</span><span class="n">mRNA_cd4</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
<span class="n">valid_genes</span> <span class="o">=</span> <span class="n">gene_counts</span> <span class="o">&gt;</span> <span class="mi">10</span>

<span class="n">genes</span> <span class="o">=</span> <span class="n">genes</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">valid_genes</span><span class="p">]</span>
<span class="n">mRNA_cd4</span> <span class="o">=</span> <span class="n">mRNA_cd4</span><span class="p">[</span><span class="n">valid_genes</span><span class="o">.</span><span class="n">nonzero</span><span class="p">()[</span><span class="mi">0</span><span class="p">],</span> <span class="p">:]</span>

<span class="c1"># Create a counts dataframe and filter other data structures</span>
<span class="n">counts</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">mRNA_cd4</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;float64&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">todense</span><span class="p">(),</span> <span class="n">columns</span><span class="o">=</span><span class="n">barcodes</span><span class="p">[</span><span class="n">is_cd4</span><span class="p">],</span> <span class="n">index</span><span class="o">=</span><span class="n">genes</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<span class="n">umi_counts</span> <span class="o">=</span> <span class="n">umi_counts</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">is_cd4</span><span class="p">]</span>
<span class="n">proteins</span> <span class="o">=</span> <span class="n">proteins</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">is_cd4</span><span class="p">]</span>
<span class="n">mito_percent</span> <span class="o">=</span> <span class="n">mito_percent</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">is_cd4</span><span class="p">]</span>

<span class="nb">print</span><span class="p">(</span><span class="n">counts</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(11608, 1486)
</pre></div></div>
</div>
</div>
<div class="section" id="Compute-a-Latent-Space-with-PCA">
<h3>Compute a Latent Space with PCA<a class="headerlink" href="#Compute-a-Latent-Space-with-PCA" title="Permalink to this headline">¶</a></h3>
<p>In the Hotspot manuscript, we used <a class="reference external" href="https://github.com/yoseflab/scvi">scVI</a> for this which has improved performance over PCA. However, to make this notebook easier to run (with less added extra dependencies), we’ll just run PCA using scikit-learn</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">counts_z</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span><span class="n">umi_counts</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">*</span><span class="n">umi_counts</span><span class="o">.</span><span class="n">median</span><span class="p">()</span>
<span class="n">counts_z</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">counts_z</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">counts_z</span> <span class="o">=</span> <span class="n">counts_z</span> \
    <span class="o">.</span><span class="n">subtract</span><span class="p">(</span><span class="n">counts</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> \
    <span class="o">.</span><span class="n">divide</span><span class="p">(</span><span class="n">counts</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="n">model</span> <span class="o">=</span> <span class="n">PCA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>
<span class="n">model</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">counts_z</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>

<span class="n">N_COMP</span> <span class="o">=</span> <span class="mi">5</span>

<span class="n">pca_data</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">counts_z</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">T</span><span class="p">)[:,</span> <span class="mi">0</span><span class="p">:</span><span class="n">N_COMP</span><span class="p">]</span>
<span class="n">pca_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">pca_data</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">counts</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">explained_variance_ratio_</span><span class="p">,</span> <span class="s1">&#39;o-&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">vlines</span><span class="p">([</span><span class="n">N_COMP</span><span class="p">],</span> <span class="o">-</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;orange&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">model</span><span class="o">.</span><span class="n">explained_variance_ratio_</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="o">*</span><span class="mf">1.1</span><span class="p">)</span>

<span class="n">pca_data</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>0</th>
      <th>1</th>
      <th>2</th>
      <th>3</th>
      <th>4</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AAACCCAAGGCCTAGA-1</th>
      <td>-5.078668</td>
      <td>5.228329</td>
      <td>-5.669723</td>
      <td>-1.739015</td>
      <td>4.327831</td>
    </tr>
    <tr>
      <th>AAACCCAGTCGTGCCA-1</th>
      <td>7.523724</td>
      <td>-6.016051</td>
      <td>-2.130812</td>
      <td>2.785405</td>
      <td>2.296527</td>
    </tr>
    <tr>
      <th>AAACCCATCGTGCATA-1</th>
      <td>6.592712</td>
      <td>-5.384366</td>
      <td>-2.021500</td>
      <td>-3.544320</td>
      <td>-4.352100</td>
    </tr>
    <tr>
      <th>AAACGAAGTAATGATG-1</th>
      <td>12.634856</td>
      <td>1.586387</td>
      <td>5.746677</td>
      <td>8.439582</td>
      <td>8.213588</td>
    </tr>
    <tr>
      <th>AAACGCTCATGCACTA-1</th>
      <td>-5.118548</td>
      <td>2.678905</td>
      <td>-0.658709</td>
      <td>0.258504</td>
      <td>1.075994</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/CD4_Tutorial_8_1.png" src="_images/CD4_Tutorial_8_1.png" />
</div>
</div>
</div>
</div>
<div class="section" id="Creating-the-Hotspot-object">
<h2>Creating the Hotspot object<a class="headerlink" href="#Creating-the-Hotspot-object" title="Permalink to this headline">¶</a></h2>
<p>To start an analysis, first create the hotspot object When creating the object, you need to specify:</p>
<ul class="simple">
<li>The gene counts matrix</li>
<li>Which background model to use</li>
<li>The latent space we are using to compute our cell metric<ul>
<li>Here we use the reduced, PCA results</li>
</ul>
</li>
<li>The per-cell scaling factor<ul>
<li>Here we use the number of umi per barcode</li>
</ul>
</li>
</ul>
<p>In this case, only the <em>log-normalized counts</em> are made available. In the Hotspot publication, we used the raw counts and the negative binomial (‘danb’) model. However, to make this example easier to run, here we just supply the log-normalized counts and select the ‘normal’ model.</p>
<p>Once the object is created, the neighborhood is then computed with <code class="docutils literal notranslate"><span class="pre">create_knn_graph</span></code></p>
<p>The two options that are specificied are <code class="docutils literal notranslate"><span class="pre">n_neighbors</span></code> which determines the size of the neighborhood, and <code class="docutils literal notranslate"><span class="pre">weighted_graph</span></code>.</p>
<p>Here we set <code class="docutils literal notranslate"><span class="pre">weighted_graph=False</span></code> to just use binary, 0-1 weights (only binary weights are supported when using a lineage tree) and <code class="docutils literal notranslate"><span class="pre">n_neighbors=30</span></code> to create a local neighborhood size of the nearest 30 cells. Larger neighborhood sizes can result in more robust detection of correlations and autocorrelations at a cost of missing more fine-grained, smaller-scale patterns.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Create the Hotspot object and the neighborhood graph</span>

<span class="n">hs</span> <span class="o">=</span> <span class="n">hotspot</span><span class="o">.</span><span class="n">Hotspot</span><span class="p">(</span><span class="n">counts</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="s1">&#39;danb&#39;</span><span class="p">,</span> <span class="n">latent</span><span class="o">=</span><span class="n">pca_data</span><span class="p">,</span> <span class="n">umi_counts</span><span class="o">=</span><span class="n">umi_counts</span><span class="p">)</span>

<span class="n">hs</span><span class="o">.</span><span class="n">create_knn_graph</span><span class="p">(</span>
    <span class="n">weighted_graph</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Determining-informative-genes">
<h2>Determining informative genes<a class="headerlink" href="#Determining-informative-genes" title="Permalink to this headline">¶</a></h2>
<p>Now we compute autocorrelations for each gene, in the pca-space, to determine which genes have the most informative variation.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">hs_results</span> <span class="o">=</span> <span class="n">hs</span><span class="o">.</span><span class="n">compute_autocorrelations</span><span class="p">(</span><span class="n">jobs</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>

<span class="n">hs_results</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 11608/11608 [00:02&lt;00:00, 5302.93it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>C</th>
      <th>Z</th>
      <th>Pval</th>
      <th>FDR</th>
    </tr>
    <tr>
      <th>Gene</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000100450</th>
      <td>0.627850</td>
      <td>117.356113</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>ENSG00000105374</th>
      <td>0.710585</td>
      <td>104.588424</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>ENSG00000166710</th>
      <td>0.580093</td>
      <td>103.624232</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>ENSG00000145425</th>
      <td>0.615415</td>
      <td>95.902552</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>ENSG00000271503</th>
      <td>0.544604</td>
      <td>93.774720</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>ENSG00000196154</th>
      <td>0.679213</td>
      <td>89.212750</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>ENSG00000145649</th>
      <td>0.518453</td>
      <td>83.451278</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>ENSG00000142676</th>
      <td>0.542604</td>
      <td>79.965974</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>ENSG00000186468</th>
      <td>0.515151</td>
      <td>79.432114</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>ENSG00000144713</th>
      <td>0.564367</td>
      <td>78.765036</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>ENSG00000110700</th>
      <td>0.486095</td>
      <td>78.423143</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>ENSG00000137441</th>
      <td>0.549342</td>
      <td>77.677710</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>ENSG00000137154</th>
      <td>0.510496</td>
      <td>75.949486</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>ENSG00000163682</th>
      <td>0.546237</td>
      <td>75.687056</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>ENSG00000156482</th>
      <td>0.572890</td>
      <td>75.220909</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</div>
<div class="section" id="Grouping-genes-into-lineage-based-modules">
<h2>Grouping genes into lineage-based modules<a class="headerlink" href="#Grouping-genes-into-lineage-based-modules" title="Permalink to this headline">¶</a></h2>
<p>To get a better idea of what expression patterns exist, it is helpful to group the genes into modules.</p>
<p>Hotspot does this using the concept of “local correlations” - that is, correlations that are computed between genes between cells in the same neighborhood.</p>
<p>Here we avoid running the calculation for all Genes x Genes pairs and instead only run this on genes that have significant autocorrelation to begin with.</p>
<p>The method <code class="docutils literal notranslate"><span class="pre">compute_local_correlations</span></code> returns a Genes x Genes matrix of Z-scores for the significance of the correlation between genes. This object is also retained in the <code class="docutils literal notranslate"><span class="pre">hs</span></code> object and is used in the subsequent steps.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Select the genes with significant lineage autocorrelation</span>
<span class="n">hs_genes</span> <span class="o">=</span> <span class="n">hs_results</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">hs_results</span><span class="o">.</span><span class="n">FDR</span> <span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;Z&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">500</span><span class="p">)</span><span class="o">.</span><span class="n">index</span>

<span class="c1"># Compute pair-wise local correlations between these genes</span>
<span class="n">lcz</span> <span class="o">=</span> <span class="n">hs</span><span class="o">.</span><span class="n">compute_local_correlations</span><span class="p">(</span><span class="n">hs_genes</span><span class="p">,</span> <span class="n">jobs</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 500/500 [00:00&lt;00:00, 6036.72it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Computing pair-wise local correlation on 500 features...
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

100%|██████████| 124750/124750 [00:06&lt;00:00, 19408.02it/s]
</pre></div></div>
</div>
<p>Now that pair-wise local correlations are calculated, we can group genes into modules.</p>
<p>To do this, a convenience method is included <code class="docutils literal notranslate"><span class="pre">create_modules</span></code> which performs agglomerative clustering with two caveats:</p>
<ul class="simple">
<li>If the FDR-adjusted p-value of the correlation between two branches exceeds <code class="docutils literal notranslate"><span class="pre">fdr_threshold</span></code>, then the branches are not merged.</li>
<li>If two branches are two be merged and they are both have at least <code class="docutils literal notranslate"><span class="pre">min_gene_threshold</span></code> genes, then the branches are not merged. Further genes that would join to the resulting merged module (and are therefore ambiguous) either remain unassigned (if <code class="docutils literal notranslate"><span class="pre">core_only=True</span></code>) or are assigned to the module with the smaller average correlations between genes, i.e.&nbsp;the least-dense module (if <code class="docutils literal notranslate"><span class="pre">core_only=False</span></code>)</li>
</ul>
<p>The output is a Series that maps gene to module number. Unassigned genes are indicated with a module number of <code class="docutils literal notranslate"><span class="pre">-1</span></code></p>
<p>This method was used to preserved substructure (nested modules) while still giving the analyst some control. However, since there are a lot of ways to do hierarchical clustering, you can also manually cluster using the gene-distances in <code class="docutils literal notranslate"><span class="pre">hs.local_correlation_z</span></code></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">modules</span> <span class="o">=</span> <span class="n">hs</span><span class="o">.</span><span class="n">create_modules</span><span class="p">(</span>
    <span class="n">min_gene_threshold</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">core_only</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fdr_threshold</span><span class="o">=</span><span class="mf">0.05</span>
<span class="p">)</span>

<span class="n">modules</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
-1     102
 1      84
 3      50
 2      46
 7      38
 8      28
 6      28
 11     26
 10     25
 9      21
 5      21
 4      16
 12     15
Name: Module, dtype: int64
</pre></div></div>
</div>
</div>
<div class="section" id="Plotting-module-correlations">
<h2>Plotting module correlations<a class="headerlink" href="#Plotting-module-correlations" title="Permalink to this headline">¶</a></h2>
<p>A convenience method is supplied to plot the results of <code class="docutils literal notranslate"><span class="pre">hs.create_modules</span></code></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">hs</span><span class="o">.</span><span class="n">plot_local_correlations</span><span class="p">(</span><span class="n">vmin</span><span class="o">=-</span><span class="mi">12</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/CD4_Tutorial_18_0.png" src="_images/CD4_Tutorial_18_0.png" />
</div>
</div>
<p>To explore individual genes, we can look at the genes with the top <em>autocorrelation</em> in a given module as these are likely the most informative.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Show the top genes for a module</span>

<span class="n">module</span> <span class="o">=</span> <span class="mi">9</span>

<span class="n">results</span> <span class="o">=</span> <span class="n">hs</span><span class="o">.</span><span class="n">results</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">hs</span><span class="o">.</span><span class="n">modules</span><span class="p">)</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">results</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">results</span><span class="o">.</span><span class="n">Module</span> <span class="o">==</span> <span class="n">module</span><span class="p">]</span>

<span class="n">genes</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">results</span><span class="p">)</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;Z&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>GeneSymbol</th>
      <th>C</th>
      <th>Z</th>
      <th>Pval</th>
      <th>FDR</th>
      <th>Module</th>
    </tr>
    <tr>
      <th>Id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000111348</th>
      <td>ARHGDIB</td>
      <td>0.190166</td>
      <td>19.955840</td>
      <td>6.668110e-89</td>
      <td>3.225142e-87</td>
      <td>9.0</td>
    </tr>
    <tr>
      <th>ENSG00000172965</th>
      <td>MIR4435-2HG</td>
      <td>0.213683</td>
      <td>18.839304</td>
      <td>1.798315e-79</td>
      <td>8.091022e-78</td>
      <td>9.0</td>
    </tr>
    <tr>
      <th>ENSG00000130592</th>
      <td>LSP1</td>
      <td>0.146799</td>
      <td>17.497352</td>
      <td>7.504333e-69</td>
      <td>3.045815e-67</td>
      <td>9.0</td>
    </tr>
    <tr>
      <th>ENSG00000163041</th>
      <td>H3F3A</td>
      <td>0.209711</td>
      <td>16.821072</td>
      <td>8.551819e-64</td>
      <td>3.320051e-62</td>
      <td>9.0</td>
    </tr>
    <tr>
      <th>ENSG00000175567</th>
      <td>UCP2</td>
      <td>0.145305</td>
      <td>16.110223</td>
      <td>1.081210e-58</td>
      <td>3.984346e-57</td>
      <td>9.0</td>
    </tr>
    <tr>
      <th>ENSG00000128340</th>
      <td>RAC2</td>
      <td>0.109768</td>
      <td>16.004655</td>
      <td>5.928431e-58</td>
      <td>2.164064e-56</td>
      <td>9.0</td>
    </tr>
    <tr>
      <th>ENSG00000102879</th>
      <td>CORO1A</td>
      <td>0.126065</td>
      <td>15.625450</td>
      <td>2.442300e-55</td>
      <td>8.723145e-54</td>
      <td>9.0</td>
    </tr>
    <tr>
      <th>ENSG00000155368</th>
      <td>DBI</td>
      <td>0.108911</td>
      <td>14.900984</td>
      <td>1.623814e-50</td>
      <td>5.479428e-49</td>
      <td>9.0</td>
    </tr>
    <tr>
      <th>ENSG00000113387</th>
      <td>SUB1</td>
      <td>0.127193</td>
      <td>14.569997</td>
      <td>2.179551e-48</td>
      <td>7.187565e-47</td>
      <td>9.0</td>
    </tr>
    <tr>
      <th>ENSG00000166913</th>
      <td>YWHAB</td>
      <td>0.095454</td>
      <td>13.553244</td>
      <td>3.790254e-42</td>
      <td>1.154784e-40</td>
      <td>9.0</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</div>
<div class="section" id="Summary-Module-Scores">
<h2>Summary Module Scores<a class="headerlink" href="#Summary-Module-Scores" title="Permalink to this headline">¶</a></h2>
<p>To aid in the recognition of the general behavior of a module, Hotspot can compute aggregate module scores.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">module_scores</span> <span class="o">=</span> <span class="n">hs</span><span class="o">.</span><span class="n">calculate_module_scores</span><span class="p">()</span>

<span class="n">module_scores</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
  0%|          | 0/12 [00:00&lt;?, ?it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Computing scores for 12 modules...
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 12/12 [00:00&lt;00:00, 28.91it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>1</th>
      <th>2</th>
      <th>3</th>
      <th>4</th>
      <th>5</th>
      <th>6</th>
      <th>7</th>
      <th>8</th>
      <th>9</th>
      <th>10</th>
      <th>11</th>
      <th>12</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AAACCCAAGGCCTAGA-1</th>
      <td>0.865375</td>
      <td>-2.542730</td>
      <td>-1.345464</td>
      <td>-0.719308</td>
      <td>-0.211530</td>
      <td>2.787655</td>
      <td>-2.001294</td>
      <td>-0.901303</td>
      <td>-0.016883</td>
      <td>-0.010592</td>
      <td>-1.065178</td>
      <td>-0.573967</td>
    </tr>
    <tr>
      <th>AAACCCAGTCGTGCCA-1</th>
      <td>-5.822828</td>
      <td>5.291330</td>
      <td>6.893498</td>
      <td>1.929916</td>
      <td>0.701222</td>
      <td>-2.890965</td>
      <td>4.678265</td>
      <td>0.784604</td>
      <td>1.677809</td>
      <td>-0.364284</td>
      <td>-0.043380</td>
      <td>0.914104</td>
    </tr>
    <tr>
      <th>AAACCCATCGTGCATA-1</th>
      <td>-2.320981</td>
      <td>3.718257</td>
      <td>0.163686</td>
      <td>1.792634</td>
      <td>-0.244096</td>
      <td>-2.512566</td>
      <td>1.363137</td>
      <td>1.236101</td>
      <td>1.207091</td>
      <td>-0.223936</td>
      <td>3.266044</td>
      <td>-0.092414</td>
    </tr>
    <tr>
      <th>AAACGAAGTAATGATG-1</th>
      <td>-8.642907</td>
      <td>4.188743</td>
      <td>6.984725</td>
      <td>1.253119</td>
      <td>0.875582</td>
      <td>-2.665184</td>
      <td>4.113193</td>
      <td>1.329732</td>
      <td>0.570870</td>
      <td>0.624170</td>
      <td>-0.132901</td>
      <td>2.513368</td>
    </tr>
    <tr>
      <th>AAACGCTCATGCACTA-1</th>
      <td>2.317936</td>
      <td>-2.531565</td>
      <td>-1.448488</td>
      <td>-1.054099</td>
      <td>-0.382518</td>
      <td>1.887807</td>
      <td>-1.614370</td>
      <td>-1.104131</td>
      <td>-0.870459</td>
      <td>-0.282801</td>
      <td>-1.141042</td>
      <td>-0.547763</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Here we can visualize these module scores by plotting them over a UMAP of the cells</p>
<p>First we’ll compute the UMAP</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">umap_data</span> <span class="o">=</span> <span class="n">UMAP</span><span class="p">(</span><span class="n">n_neighbors</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">min_dist</span><span class="o">=.</span><span class="mi">5</span><span class="p">)</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">pca_data</span><span class="p">)</span>
<span class="n">umap_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">umap_data</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">counts</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;UMAP1&#39;</span><span class="p">,</span> <span class="s1">&#39;UMAP2&#39;</span><span class="p">])</span>

<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">umap_data</span><span class="o">.</span><span class="n">UMAP1</span><span class="p">,</span> <span class="n">umap_data</span><span class="o">.</span><span class="n">UMAP2</span><span class="p">,</span> <span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">ms</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;UMAP-1&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;UMAP-2&#39;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">sp</span> <span class="ow">in</span> <span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span><span class="o">.</span><span class="n">spines</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
    <span class="n">sp</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">([])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">([]);</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/CD4_Tutorial_24_0.png" src="_images/CD4_Tutorial_24_0.png" />
</div>
</div>
<p>Finally, we plot the module scores on top of the UMAP for comparison</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="n">cax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_axes</span><span class="p">(</span>
    <span class="p">[</span><span class="o">.</span><span class="mi">95</span><span class="p">,</span> <span class="o">.</span><span class="mi">15</span><span class="p">,</span> <span class="o">.</span><span class="mi">007</span><span class="p">,</span> <span class="o">.</span><span class="mi">1</span><span class="p">]</span>
<span class="p">)</span>

<span class="k">for</span> <span class="n">ax</span><span class="p">,</span> <span class="n">mod</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">axs</span><span class="o">.</span><span class="n">ravel</span><span class="p">(),</span> <span class="n">hs</span><span class="o">.</span><span class="n">module_scores</span><span class="o">.</span><span class="n">columns</span><span class="p">):</span>
    <span class="n">sc</span> <span class="o">=</span> <span class="n">hs</span><span class="o">.</span><span class="n">module_scores</span><span class="p">[</span><span class="n">mod</span><span class="p">]</span>
    <span class="c1"># vmin = np.percentile(sc, 5)</span>
    <span class="c1"># vmax = np.percentile(sc, 95)</span>
    <span class="n">vmin</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
    <span class="n">vmax</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">sca</span><span class="p">(</span><span class="n">ax</span><span class="p">)</span>
    <span class="n">sc</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
        <span class="n">umap_data</span><span class="o">.</span><span class="n">UMAP1</span><span class="p">,</span> <span class="n">umap_data</span><span class="o">.</span><span class="n">UMAP2</span><span class="p">,</span>
        <span class="n">s</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="n">sc</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="n">vmin</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="n">vmax</span><span class="p">,</span>
        <span class="n">rasterized</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">([])</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">([])</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Module </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">mod</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">sp</span> <span class="ow">in</span> <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="n">sp</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">subplots_adjust</span><span class="p">(</span><span class="n">hspace</span><span class="o">=</span><span class="mf">0.4</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">sc</span><span class="p">,</span> <span class="n">cax</span><span class="o">=</span><span class="n">cax</span><span class="p">,</span> <span class="n">ticks</span><span class="o">=</span><span class="p">[</span><span class="n">vmin</span><span class="p">,</span> <span class="n">vmax</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Module</span><span class="se">\n</span><span class="s1">Score&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">subplots_adjust</span><span class="p">(</span><span class="n">left</span><span class="o">=</span><span class="mf">0.02</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="mf">0.9</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/CD4_Tutorial_26_0.png" src="_images/CD4_Tutorial_26_0.png" />
</div>
</div>
</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2020, David DeTomaso.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.2.<br/>
    </p>
  </div>
</footer>
  </body>
</html>