<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Demo: Spatial data from Slide-seq &#8212; Hotspot 0.9.0 documentation</title>
    <link rel="stylesheet" href="_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/my-styles.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Demo: Lineage-recorded data from Chan et al, (Nature 2019)" href="Lineage_Tutorial.html" />
    <link rel="prev" title="Hotspot" href="index.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<script type="text/javascript" src="_static/js/jquery-1.11.0.min.js "></script>
<script type="text/javascript" src="_static/js/jquery-fix.js "></script>
<script type="text/javascript" src="_static/bootstrap-3.3.7/js/bootstrap.min.js "></script>
<script type="text/javascript" src="_static/bootstrap-sphinx.js "></script>

  </head><body>

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="index.html">
          Hotspot</a>
        <span class="navbar-text navbar-version pull-left"><b>0.9.0</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="hotspot.html">API</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="index.html">Tutorials <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Demo: Spatial data from Slide-seq</a></li>
<li class="toctree-l1"><a class="reference internal" href="Lineage_Tutorial.html">Demo: Lineage-recorded data from Chan et al, (Nature 2019)</a></li>
<li class="toctree-l1"><a class="reference internal" href="CD4_Tutorial.html">Demo: Transcriptional Patterns in CD4 T Cells</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">This Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">Demo: Spatial data from Slide-seq</a><ul>
<li><a class="reference internal" href="#Covered-Here">Covered Here</a></li>
<li><a class="reference internal" href="#Creating-the-Hotspot-object">Creating the Hotspot object</a></li>
<li><a class="reference internal" href="#Determining-genes-with-spatial-variation">Determining genes with spatial variation</a></li>
<li><a class="reference internal" href="#Grouping-genes-into-spatial-modules">Grouping genes into spatial modules</a></li>
<li><a class="reference internal" href="#Plotting-module-correlations">Plotting module correlations</a></li>
<li><a class="reference internal" href="#Summary-Module-Scores">Summary Module Scores</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="body col-md-12 content" role="main">
      
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container,
div.nbinput.container div.prompt,
div.nbinput.container div.input_area,
div.nbinput.container div[class*=highlight],
div.nbinput.container div[class*=highlight] pre,
div.nboutput.container,
div.nboutput.container div.prompt,
div.nboutput.container div.output_area,
div.nboutput.container div[class*=highlight],
div.nboutput.container div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    min-width: 7ex;
    padding-top: 0.4em;
    padding-right: 0.4em;
    text-align: right;
    flex: 0;
}
@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    padding: 0.4em;
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    background: #f5f5f5;
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<div class="section" id="Demo:-Spatial-data-from-Slide-seq">
<h1>Demo: Spatial data from Slide-seq<a class="headerlink" href="#Demo:-Spatial-data-from-Slide-seq" title="Permalink to this headline">¶</a></h1>
<p>This demo uses data from:</p>
<p><em>Slide-seq: A scalable technology for measuring genome-wide expression at high spatial resolution</em></p>
<ul class="simple">
<li>Rodriques et al (Science 2019)</li>
<li><a class="reference external" href="https://doi.org/10.1126/science.aaw1219">https://doi.org/10.1126/science.aaw1219</a></li>
</ul>
<p>For this demo, you need the data files for Puck_180819_12</p>
<p>Taken from: <a class="reference external" href="https://portals.broadinstitute.org/single_cell/study/SCP354/slide-seq-study#study-summary">https://portals.broadinstitute.org/single_cell/study/SCP354/slide-seq-study#study-summary</a></p>
<p>Specifically these two files:</p>
<ul class="simple">
<li>MappedDGEForR.csv</li>
<li>BeadMapping_10-17_1014/BeadLocationsForR.csv</li>
</ul>
<div class="section" id="Covered-Here">
<h2>Covered Here<a class="headerlink" href="#Covered-Here" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>Data preprocessing and filtering</li>
<li>Computing Autocorrelation in Hotspot to identify spatial genes</li>
<li>Computing local correlations between spatial genes to identify modules</li>
<li>Plotting modules, correlations, and module scores</li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">hotspot</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">matplotlib.colors</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>

<span class="c1"># Load the counts and positions</span>
<span class="n">counts_file</span> <span class="o">=</span> <span class="s1">&#39;MappedDGEForR.csv&#39;</span>
<span class="n">pos_file</span> <span class="o">=</span> <span class="s1">&#39;BeadMapping_10-17_1014/BeadLocationsForR.csv&#39;</span>

<span class="n">pos</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">pos_file</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">counts</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">counts_file</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="c1"># Takes a while, ~10min</span>

<span class="c1"># Align the indices</span>
<span class="n">counts</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">pos</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
<span class="n">barcodes</span> <span class="o">=</span> <span class="n">pos</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span>

<span class="c1"># Swap position axes</span>
<span class="c1"># We swap x&#39;=y and y&#39;=-x to match the slides in the paper</span>
<span class="n">pos</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s1">&#39;X&#39;</span><span class="p">:</span> <span class="n">pos</span><span class="o">.</span><span class="n">ycoord</span><span class="p">,</span>
        <span class="s1">&#39;Y&#39;</span><span class="p">:</span> <span class="n">pos</span><span class="o">.</span><span class="n">xcoord</span><span class="o">*-</span><span class="mi">1</span><span class="p">,</span>
    <span class="p">},</span> <span class="n">index</span><span class="o">=</span><span class="n">pos</span><span class="o">.</span><span class="n">index</span>
<span class="p">)</span>

<span class="n">num_umi</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># Filter genes</span>
<span class="n">gene_counts</span> <span class="o">=</span> <span class="p">(</span><span class="n">counts</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">valid_genes</span> <span class="o">=</span> <span class="n">gene_counts</span> <span class="o">&gt;=</span> <span class="mi">50</span>
<span class="n">counts</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">valid_genes</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Creating-the-Hotspot-object">
<h2>Creating the Hotspot object<a class="headerlink" href="#Creating-the-Hotspot-object" title="Permalink to this headline">¶</a></h2>
<p>To start an analysis, first creat the hotspot object When creating the object, you need to specify:</p>
<ul class="simple">
<li>The gene counts matrix</li>
<li>Which background model to use</li>
<li>The latent space we are using to compute our cell metric<ul>
<li>Here we use the physical barcode positions</li>
</ul>
</li>
<li>The per-cell scaling factor<ul>
<li>Here we use the number of umi per barcode</li>
</ul>
</li>
</ul>
<p>Once the object is created, the neighborhood graph must be computed with <code class="docutils literal notranslate"><span class="pre">create_knn_graph</span></code></p>
<p>The two options that are specificied are <code class="docutils literal notranslate"><span class="pre">n_neighbors</span></code> which determines the size of the neighborhood, and <code class="docutils literal notranslate"><span class="pre">weighted_graph</span></code>.</p>
<p>Here we set <code class="docutils literal notranslate"><span class="pre">weighted_graph=False</span></code> to just use binary, 0-1 weights and <code class="docutils literal notranslate"><span class="pre">n_neighbors=300</span></code> to create a local neighborhood size of the nearest 300 barcodes. Larger neighborhood sizes can result in more robust detection of correlations and autocorrelations at a cost of missing more fine-grained, smaller-scale, spatial patterns</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Create the Hotspot object and the neighborhood graph</span>

<span class="n">hs</span> <span class="o">=</span> <span class="n">hotspot</span><span class="o">.</span><span class="n">Hotspot</span><span class="p">(</span><span class="n">counts</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="s1">&#39;bernoulli&#39;</span><span class="p">,</span> <span class="n">latent</span><span class="o">=</span><span class="n">pos</span><span class="p">,</span> <span class="n">umi_counts</span><span class="o">=</span><span class="n">num_umi</span><span class="p">)</span>

<span class="n">hs</span><span class="o">.</span><span class="n">create_knn_graph</span><span class="p">(</span>
    <span class="n">weighted_graph</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Determining-genes-with-spatial-variation">
<h2>Determining genes with spatial variation<a class="headerlink" href="#Determining-genes-with-spatial-variation" title="Permalink to this headline">¶</a></h2>
<p>Now we compute autocorrelations for each gene, using the spatial metric, to determine which genes have the most spatial variation.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">hs_results</span> <span class="o">=</span> <span class="n">hs</span><span class="o">.</span><span class="n">compute_autocorrelations</span><span class="p">(</span><span class="n">jobs</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>

<span class="n">hs_results</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 6942/6942 [00:27&lt;00:00, 250.89it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>G</th>
      <th>EG</th>
      <th>stdG</th>
      <th>Z</th>
      <th>C</th>
      <th>Pval</th>
      <th>FDR</th>
    </tr>
    <tr>
      <th>Gene</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Ttr</th>
      <td>1.895617e+06</td>
      <td>0</td>
      <td>4349.885976</td>
      <td>435.785413</td>
      <td>0.195658</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Plp1</th>
      <td>1.070512e+06</td>
      <td>0</td>
      <td>4349.885976</td>
      <td>246.101150</td>
      <td>0.109810</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Enpp2</th>
      <td>9.047060e+05</td>
      <td>0</td>
      <td>4349.885976</td>
      <td>207.983833</td>
      <td>0.099028</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Fth1</th>
      <td>9.041625e+05</td>
      <td>0</td>
      <td>4349.885976</td>
      <td>207.858890</td>
      <td>0.092616</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Cartpt</th>
      <td>7.867150e+05</td>
      <td>0</td>
      <td>4349.885976</td>
      <td>180.858760</td>
      <td>0.084189</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</div>
<div class="section" id="Grouping-genes-into-spatial-modules">
<h2>Grouping genes into spatial modules<a class="headerlink" href="#Grouping-genes-into-spatial-modules" title="Permalink to this headline">¶</a></h2>
<p>To get a better idea of what spatial patterns exist, it is helpful to group the genes into modules.</p>
<p>Hotspot does this using the concept of “local correlations” - that is, correlations that are computed between genes between cells in the same neighborhood.</p>
<p>Here we avoid running the calculation for all Genes x Genes pairs and instead only run this on genes that have significant spatial autocorrelation to begin with.</p>
<p>The method <code class="docutils literal notranslate"><span class="pre">compute_local_correlations</span></code> returns a Genes x Genes matrix of Z-scores for the significance of the correlation between genes. This object is also retained in the <code class="docutils literal notranslate"><span class="pre">hs</span></code> object and is used in the subsequent steps.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Select the genes with significant spatial autocorrelation</span>
<span class="n">hs_genes</span> <span class="o">=</span> <span class="n">hs_results</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">hs_results</span><span class="o">.</span><span class="n">FDR</span> <span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">]</span>

<span class="c1"># Compute pair-wise local correlations between these genes</span>
<span class="n">lcz</span> <span class="o">=</span> <span class="n">hs</span><span class="o">.</span><span class="n">compute_local_correlations</span><span class="p">(</span><span class="n">hs_genes</span><span class="p">,</span> <span class="n">jobs</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
  0%|          | 0/845 [00:00&lt;?, ?it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Computing pair-wise local correlation on 845 features...
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 845/845 [00:05&lt;00:00, 157.16it/s]
100%|██████████| 356590/356590 [59:29&lt;00:00, 99.89it/s]
</pre></div></div>
</div>
<p>Now that pair-wise local correlations are calculated, we can group genes into modules.</p>
<p>To do this, a convenience method is included <code class="docutils literal notranslate"><span class="pre">create_modules</span></code> which performs agglomerative clustering with two caveats:</p>
<ul class="simple">
<li>If the FDR-adjusted p-value of the correlation between two branches exceeds <code class="docutils literal notranslate"><span class="pre">fdr_threshold</span></code>, then the branches are not merged.</li>
<li>If two branches are two be merged and they are both have at least <code class="docutils literal notranslate"><span class="pre">min_gene_threshold</span></code> genes, then the branches are not merged. Further genes that would join to the resulting merged module (and are therefore ambiguous) either remain unassigned (if <code class="docutils literal notranslate"><span class="pre">core_only=True</span></code>) or are assigned to the module with the smaller average correlations between genes, i.e.&nbsp;the least-dense module (if <code class="docutils literal notranslate"><span class="pre">core_only=False</span></code>)</li>
</ul>
<p>The output is a Series that maps gene to module number. Unassigned genes are indicated with a module number of <code class="docutils literal notranslate"><span class="pre">-1</span></code></p>
<p>This method was used to preserved substructure (nested modules) while still giving the analyst some control. However, since there are a lot of ways to do hierarchical clustering, you can also manually cluster using the gene-distances in <code class="docutils literal notranslate"><span class="pre">hs.local_correlation_z</span></code></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">modules</span> <span class="o">=</span> <span class="n">hs</span><span class="o">.</span><span class="n">create_modules</span><span class="p">(</span>
    <span class="n">min_gene_threshold</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">core_only</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">fdr_threshold</span><span class="o">=</span><span class="mf">0.05</span>
<span class="p">)</span>

<span class="n">modules</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
 4    230
 3    170
-1    143
 2    102
 1    101
 5     63
 6     36
Name: Module, dtype: int64
</pre></div></div>
</div>
</div>
<div class="section" id="Plotting-module-correlations">
<h2>Plotting module correlations<a class="headerlink" href="#Plotting-module-correlations" title="Permalink to this headline">¶</a></h2>
<p>A convenience method is supplied to plot the results of <code class="docutils literal notranslate"><span class="pre">hs.create_modules</span></code></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">hs</span><span class="o">.</span><span class="n">plot_local_correlations</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Spatial_Tutorial_12_0.png" src="_images/Spatial_Tutorial_12_0.png" />
</div>
</div>
<p>To explore individual genes, we can look at the genes with the top <em>autocorrelation</em> in a given module as these are likely the most informative.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Show the top genes for a module</span>

<span class="n">module</span> <span class="o">=</span> <span class="mi">1</span>

<span class="n">results</span> <span class="o">=</span> <span class="n">hs</span><span class="o">.</span><span class="n">results</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">hs</span><span class="o">.</span><span class="n">modules</span><span class="p">)</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">results</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">results</span><span class="o">.</span><span class="n">Module</span> <span class="o">==</span> <span class="n">module</span><span class="p">]</span>
<span class="n">results</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;Z&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>G</th>
      <th>EG</th>
      <th>stdG</th>
      <th>Z</th>
      <th>C</th>
      <th>Pval</th>
      <th>FDR</th>
      <th>Module</th>
    </tr>
    <tr>
      <th>Gene</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Pcp2</th>
      <td>454045.514481</td>
      <td>0</td>
      <td>4349.885976</td>
      <td>104.381015</td>
      <td>0.046222</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>Car8</th>
      <td>384470.481798</td>
      <td>0</td>
      <td>4349.885976</td>
      <td>88.386336</td>
      <td>0.039361</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>Itpr1</th>
      <td>309992.054843</td>
      <td>0</td>
      <td>4349.885976</td>
      <td>71.264409</td>
      <td>0.031606</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>Pcp4</th>
      <td>287016.421190</td>
      <td>0</td>
      <td>4349.885976</td>
      <td>65.982516</td>
      <td>0.029496</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>Pvalb</th>
      <td>277574.727119</td>
      <td>0</td>
      <td>4349.885976</td>
      <td>63.811955</td>
      <td>0.028504</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>mt-Cytb</th>
      <td>251459.913992</td>
      <td>0</td>
      <td>4349.885976</td>
      <td>57.808392</td>
      <td>0.025916</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>Gpm6b</th>
      <td>229566.984067</td>
      <td>0</td>
      <td>4349.885976</td>
      <td>52.775403</td>
      <td>0.023639</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>mt-Nd4</th>
      <td>226437.069373</td>
      <td>0</td>
      <td>4349.885976</td>
      <td>52.055863</td>
      <td>0.023182</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>Calb1</th>
      <td>198838.797756</td>
      <td>0</td>
      <td>4349.885976</td>
      <td>45.711267</td>
      <td>0.020217</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>Aldoc</th>
      <td>187057.748608</td>
      <td>0</td>
      <td>4349.885976</td>
      <td>43.002909</td>
      <td>0.019332</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>To get an idea of what spatial pattern the module is referencing, we can plot the top module genes’ expression onto the spatial positions</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Plot the module scores on top of positions</span>

<span class="n">module</span> <span class="o">=</span> <span class="mi">1</span>

<span class="n">results</span> <span class="o">=</span> <span class="n">hs</span><span class="o">.</span><span class="n">results</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">hs</span><span class="o">.</span><span class="n">modules</span><span class="p">)</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">results</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">results</span><span class="o">.</span><span class="n">Module</span> <span class="o">==</span> <span class="n">module</span><span class="p">]</span>
<span class="n">genes</span> <span class="o">=</span> <span class="n">results</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;Z&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">6</span><span class="p">)</span><span class="o">.</span><span class="n">index</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">11</span><span class="p">,</span> <span class="mf">7.5</span><span class="p">))</span>

<span class="n">cmap</span> <span class="o">=</span> <span class="n">matplotlib</span><span class="o">.</span><span class="n">colors</span><span class="o">.</span><span class="n">LinearSegmentedColormap</span><span class="o">.</span><span class="n">from_list</span><span class="p">(</span>
    <span class="s1">&#39;grays&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;#DDDDDD&#39;</span><span class="p">,</span> <span class="s1">&#39;#000000&#39;</span><span class="p">])</span>

<span class="k">for</span> <span class="n">ax</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">axs</span><span class="o">.</span><span class="n">ravel</span><span class="p">(),</span> <span class="n">genes</span><span class="p">):</span>

    <span class="n">expression</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">hs</span><span class="o">.</span><span class="n">counts</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">/</span><span class="n">hs</span><span class="o">.</span><span class="n">umi_counts</span><span class="o">*</span><span class="mi">45</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="c1"># log-counts per 45 (median UMI/barcode)</span>

    <span class="n">vmin</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">vmax</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">expression</span><span class="p">,</span> <span class="mi">95</span><span class="p">)</span>
    <span class="n">vmax</span> <span class="o">=</span> <span class="mi">2</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">sca</span><span class="p">(</span><span class="n">ax</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">hs</span><span class="o">.</span><span class="n">latent</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span>
                <span class="n">y</span><span class="o">=</span><span class="n">hs</span><span class="o">.</span><span class="n">latent</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span>
                <span class="n">s</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                <span class="n">c</span><span class="o">=</span><span class="n">expression</span><span class="p">,</span>
                <span class="n">vmin</span><span class="o">=</span><span class="n">vmin</span><span class="p">,</span>
                <span class="n">vmax</span><span class="o">=</span><span class="n">vmax</span><span class="p">,</span>
                <span class="n">edgecolors</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">,</span>
                <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span>
               <span class="p">)</span>

    <span class="k">for</span> <span class="n">sp</span> <span class="ow">in</span> <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="n">sp</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">([])</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">([])</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Spatial_Tutorial_16_0.png" src="_images/Spatial_Tutorial_16_0.png" />
</div>
</div>
</div>
<div class="section" id="Summary-Module-Scores">
<h2>Summary Module Scores<a class="headerlink" href="#Summary-Module-Scores" title="Permalink to this headline">¶</a></h2>
<p>To aid in the recognition of the general behavior of a module, Hotspot can compute aggregate module scores.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">module_scores</span> <span class="o">=</span> <span class="n">hs</span><span class="o">.</span><span class="n">calculate_module_scores</span><span class="p">()</span>

<span class="n">module_scores</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
  0%|          | 0/6 [00:00&lt;?, ?it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Computing scores for 6 modules...
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 6/6 [01:30&lt;00:00, 12.98s/it]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>1</th>
      <th>2</th>
      <th>3</th>
      <th>4</th>
      <th>5</th>
      <th>6</th>
    </tr>
    <tr>
      <th>barcodes</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>CGTACAATTTTTTT</th>
      <td>0.822729</td>
      <td>-0.443706</td>
      <td>-0.094899</td>
      <td>-0.414559</td>
      <td>-0.198448</td>
      <td>0.180391</td>
    </tr>
    <tr>
      <th>TTCGTTATTTTTTT</th>
      <td>1.232227</td>
      <td>-0.421417</td>
      <td>-0.254999</td>
      <td>-0.472261</td>
      <td>-0.213982</td>
      <td>0.299108</td>
    </tr>
    <tr>
      <th>CAAACCAACCCCCC</th>
      <td>0.607751</td>
      <td>-0.312146</td>
      <td>-0.166309</td>
      <td>0.234819</td>
      <td>-0.328414</td>
      <td>0.367822</td>
    </tr>
    <tr>
      <th>TCTTTTCACCCCCC</th>
      <td>1.286320</td>
      <td>-0.491108</td>
      <td>-0.053740</td>
      <td>-0.736092</td>
      <td>-0.096152</td>
      <td>0.377854</td>
    </tr>
    <tr>
      <th>GGATTGAACCCCCC</th>
      <td>0.133101</td>
      <td>-0.544696</td>
      <td>-0.157040</td>
      <td>-0.326436</td>
      <td>-0.133297</td>
      <td>0.334153</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Here we can visualize these module scores by plotting them over the barcode positions:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Plot the module scores on top of positions</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">11</span><span class="p">,</span> <span class="mf">7.5</span><span class="p">))</span>

<span class="k">for</span> <span class="n">ax</span><span class="p">,</span> <span class="n">module</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">axs</span><span class="o">.</span><span class="n">ravel</span><span class="p">(),</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">hs</span><span class="o">.</span><span class="n">modules</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="o">+</span><span class="mi">1</span><span class="p">)):</span>
    <span class="n">scores</span> <span class="o">=</span> <span class="n">hs</span><span class="o">.</span><span class="n">module_scores</span><span class="p">[</span><span class="n">module</span><span class="p">]</span>

    <span class="n">vmin</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">scores</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">vmax</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">scores</span><span class="p">,</span> <span class="mi">99</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">sca</span><span class="p">(</span><span class="n">ax</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">hs</span><span class="o">.</span><span class="n">latent</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span>
                <span class="n">y</span><span class="o">=</span><span class="n">hs</span><span class="o">.</span><span class="n">latent</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span>
                <span class="n">s</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                <span class="n">c</span><span class="o">=</span><span class="n">scores</span><span class="p">,</span>
                <span class="n">vmin</span><span class="o">=</span><span class="n">vmin</span><span class="p">,</span>
                <span class="n">vmax</span><span class="o">=</span><span class="n">vmax</span><span class="p">,</span>
                <span class="n">edgecolors</span><span class="o">=</span><span class="s1">&#39;none&#39;</span>
               <span class="p">)</span>

    <span class="k">for</span> <span class="n">sp</span> <span class="ow">in</span> <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="n">sp</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">([])</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">([])</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Module </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">module</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Spatial_Tutorial_20_0.png" src="_images/Spatial_Tutorial_20_0.png" />
</div>
</div>
</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2020, David DeTomaso.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.2.<br/>
    </p>
  </div>
</footer>
  </body>
</html>